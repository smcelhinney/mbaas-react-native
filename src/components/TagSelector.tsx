import * as React from 'react';
import { Component } from 'react';
import {
  Text,
  View,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  ScrollView,
  Dimensions,
  Modal,
  FlatList,
  LayoutAnimation,
  TextStyle,
  ViewStyle
} from 'react-native';
import { sharedTypographyStyles, colorScheme } from '../constants/styles';
import SuperPicker from './SuperPicker';
import { format } from 'date-fns';
import { CheckBox } from 'native-base';
import Icon from 'react-native-vector-icons/Ionicons';
import * as _ from 'lodash';

class TagSelector extends Component<any, any> {
  constructor(props: any) {
    super(props);

    this.state = {
      tags: [],
      prescriptionLengths: props.prescriptionLengths
    };
  }

  public componentWillReceiveProps(nextProps: any) {
    // LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
    this.setState({
      tags: nextProps.selectedItems,
      prescriptionLengths: nextProps.prescriptionLengths
    });
  }

  public componentDidMount() {
    this.setState({
      tags: this.props.selectedItems
    });
  }

  private updatePrescriptionLength(id: string, prescriptionLength: number) {
    const tags = this.state.tags.map((edge: any) => {
      const updatedEdge = {
        ...edge,
        prescriptionLength: prescriptionLength
      };
      return edge.node.id === id ? updatedEdge : edge;
    });

    this.setState({
      ...this.state,
      tags: tags
    });

    this.props.onPrescriptionLengthChange(prescriptionLength, id);
  }

  private updateDate(id: string, date: string) {
    const tags = this.state.tags.map((edge: any) => {
      const updatedEdge = {
        ...edge,
        date: date
      };
      return edge.node.id === id ? updatedEdge : edge;
    });

    this.setState({
      ...this.state,
      tags: tags
    });

    this.props.onDateChange(date, id);
  }

  private renderTags(tags: any) {
    if (tags.length === 0) {
      return (
        <TouchableOpacity
          onPress={() => this.props.openSelectorModal()}
          style={styles.tag}
        >
          <Text style={styles.tagText}>
            + Add {this.props.name.toLowerCase()}
          </Text>
        </TouchableOpacity>
      );
    }

    return this.state.tags.map(
      (edge: {
        node: { id: string; name: string; isAlert?: boolean };
        date?: string;
        prescriptionLength?: number;
      }) => {
        const showAlertIcon = edge.node.isAlert ? (
          <Icon name="md-warning" color="white" style={{ marginRight: 5 }} />
        ) : null;
        const alertTagStyles: ViewStyle | any = edge.node.isAlert
          ? { backgroundColor: '#F24B35' }
          : null;
        const alertTextStyles: TextStyle | any = edge.node.isAlert
          ? { color: 'white' }
          : null;

        return this.props.withDate ? (
          <TouchableOpacity
            key={edge.node.id}
            onLongPress={() => this.props.openDateSelector()}
            onPress={() => this.props.openSelectorModal()}
            style={[styles.tag, alertTagStyles, { width: '100%' }]}
          >
            {showAlertIcon}
            <View style={{ flexDirection: 'row' }}>
              <Text style={styles.tagDateText}>
                {this.props.dateText + ': '}
              </Text>
              <SuperPicker
                mode="date"
                displayFormat="DD MMM YYYY"
                inputTextStyle={[
                  styles.tagDateText,
                  { color: colorScheme.mediumOrange }
                ]}
                selectedValue={edge.date}
                onValueChange={(date: string) =>
                  this.updateDate(edge.node.id, date)
                }
              />
            </View>
            {this.props.withPrescriptionLength ? (
              <View style={{ flexDirection: 'row' }}>
                <Text style={styles.tagDateText}>
                  {this.props.prescriptionLengthText + ': '}
                </Text>
                <SuperPicker
                  mode="select"
                  onValueChange={(length: number) =>
                    this.updatePrescriptionLength(edge.node.id, length)
                  }
                  options={this.state.prescriptionLengths}
                  inputTextStyle={[
                    styles.tagDateText,
                    { color: colorScheme.mediumOrange }
                  ]}
                  selectedValue={edge.prescriptionLength}
                />
              </View>
            ) : null}
            <Text style={[styles.tagText, alertTextStyles]}>
              {edge.node.name}
            </Text>
          </TouchableOpacity>
        ) : (
          <TouchableOpacity
            onPress={() => this.props.openSelectorModal()}
            key={edge.node.id}
            style={[styles.tag, alertTagStyles]}
          >
            {showAlertIcon}
            <Text style={[styles.tagText, alertTextStyles]}>
              {edge.node.name}
            </Text>
          </TouchableOpacity>
        );
      }
    );
  }

  public render() {
    const { selectedItems, name, color, count, withDate } = this.props;

    return (
      <View style={styles.primaryListSection}>
        <View style={[styles.colorCodeBand, { backgroundColor: color }]} />
        <View style={styles.primaryListSectionContent}>
          <Text style={styles.cardSubHeaderText}>
            {count + ' ' + name.toUpperCase()}
          </Text>
          <View style={styles.tagContainer}>
            {this.renderTags(this.state.tags)}
          </View>
        </View>
      </View>
    );
  }
}

export default TagSelector;

const styles = StyleSheet.create({
  ...sharedTypographyStyles,
  primaryListSection: {
    flexDirection: 'row',
    backgroundColor: colorScheme.white,
    // borderBottomWidth: StyleSheet.hairlineWidth,
    borderRadius: 5,
    borderColor: '#CCCCCC',
    margin: 10,
    shadowRadius: 10,
    shadowOpacity: 0.1,
    shadowColor: 'black'
  },
  primaryListSectionHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  primaryListSectionContent: {
    flex: 1,
    paddingVertical: 15,
    paddingHorizontal: 10
  },
  colorCodeBand: {
    alignItems: 'center',
    width: 8,
    paddingTop: 10,
    borderTopLeftRadius: 5,
    borderBottomLeftRadius: 5
  },
  countText: {
    fontSize: 18,
    fontWeight: 'bold',
    color: colorScheme.white
  },
  tagContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    flexWrap: 'wrap'
  },
  tag: {
    backgroundColor: '#EDF2FF',
    paddingHorizontal: 10,
    paddingVertical: 7,
    borderRadius: 8,
    alignItems: 'flex-start',
    justifyContent: 'center',
    marginRight: 3,
    marginBottom: 3
  },
  checkBoxWrapper: {
    width: 30
  },
  modalContainer: {
    flex: 1
  },
  modalHeader: {
    marginTop: 20,
    alignItems: 'flex-end',
    borderBottomWidth: 1,
    borderColor: '#CCCCCC'
  },
  modalCloseButton: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 40,
    marginRight: 10
  },
  listItem: {
    flexDirection: 'row',
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderColor: '#CCCCCC',
    height: 40,
    paddingLeft: 15,
    alignItems: 'center'
  },
  listItemText: {
    paddingLeft: 15,
    fontFamily: 'Helvetica Neue',
    fontSize: 12,
    color: '#333333'
  }
});
