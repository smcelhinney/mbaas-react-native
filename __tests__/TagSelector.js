import 'react-native';
import React from 'react';
import TagSelector from '../build/components/TagSelector.js';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const minProps = {
  selectedItems: [
    {
      node: {
        id: '01',
        isAlert: false,
        name: 'TEST_TAG'
      }
    }
  ]
}

describe('Component: TagSelector', () => {

  it('TagSelector renders correctly', () => {
    const wrapper = shallow(
      <TagSelector name="NAME" {...minProps}/>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  })

});