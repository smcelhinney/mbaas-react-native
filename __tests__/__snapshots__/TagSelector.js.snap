// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TagSelector TagSelector renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#CCCCCC",
      "borderRadius": 5,
      "flexDirection": "row",
      "margin": 10,
      "shadowColor": "black",
      "shadowOpacity": 0.1,
      "shadowRadius": 10,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomLeftRadius": 5,
          "borderTopLeftRadius": 5,
          "paddingTop": 10,
          "width": 8,
        },
        Object {
          "backgroundColor": undefined,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#8e8e93",
          "fontFamily": "Helvetica Neue",
          "fontSize": 11,
          "fontWeight": "bold",
          "marginBottom": 7,
        }
      }
    >
      undefined NAME
    </Text>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        key="01"
        onPress={[Function]}
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "backgroundColor": "#EDF2FF",
              "borderRadius": 8,
              "justifyContent": "center",
              "marginBottom": 3,
              "marginRight": 3,
              "paddingHorizontal": 10,
              "paddingVertical": 7,
            },
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#555555",
                "fontSize": 14,
              },
              null,
            ]
          }
        >
          TEST_TAG
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
